OpenSim/Simulation/SimulationUtilities.cpp

```diff
@@ -0,0 +1,174 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "SimulationUtilities.h"
+
+using namespace OpenSim;
+
+SimTK::State OpenSim::simulate(Model& model,
+    const SimTK::State& initialState,
+    double finalTime,
+    bool saveStatesFile)
+{
+    // Returned state begins as a copy of the initial state
+    SimTK::State state = initialState;
+    SimTK::Visualizer::InputSilo* silo;
+
+    bool simulateOnce = true;
+
+    // Ensure the final time is in the future.
+    const double initialTime = initialState.getTime();
+    if (finalTime <= initialTime) {
+        std::cout << "The final time must be in the future (current time is "
+                  << initialTime << "); simulation aborted." << std::endl;
+        return state;
+    }
+
+    // Configure the visualizer.
+    if (model.getUseVisualizer()) {
+        SimTK::Visualizer& viz = model.updVisualizer().updSimbodyVisualizer();
+        // We use the input silo to get key presses.
+        silo = &model.updVisualizer().updInputSilo();
+
+        SimTK::DecorativeText help("Press any key to start a new simulation; "
+            "ESC to quit.");
+        help.setIsScreenText(true);
+        viz.addDecoration(SimTK::MobilizedBodyIndex(0), SimTK::Vec3(0), help);
+
+        viz.setShowSimTime(true);
+        viz.drawFrameNow(state);
+        std::cout << "A visualizer window has opened." << std::endl;
+
+        // if visualizing enable replay
+        simulateOnce = false;
+    }
+
+    // Simulate until the user presses ESC (or enters 'q' if visualization has
+    // been disabled).
+    do {
+        if (model.getUseVisualizer()) {
+            // Get a key press.
+            silo->clear(); // Ignore any previous key presses.
+            unsigned key, modifiers;
+            silo->waitForKeyHit(key, modifiers);
+            if (key == SimTK::Visualizer::InputListener::KeyEsc) { break; }
+        }
+
+        // reset the state to the initial state
+        state = initialState;
+        // Set up manager and simulate.
+        Manager manager(model);
+        state.setTime(initialTime);
+        manager.initialize(state);
+        state = manager.integrate(finalTime);
+
+        // Save the states to a storage file (if requested).
+        if (saveStatesFile) {
+            manager.getStateStorage().print(model.getName() + "_states.sto");
+        }
+    } while (!simulateOnce);
+
+    return state;
+}
+/// @}
+
+std::unique_ptr<Storage>
+OpenSim::updatePre40KinematicsStorageFor40MotionType(const Model& pre40Model,
+        const Storage &kinematics)
+{
+    // There is no issue if the kinematics are in internal values (i.e. not
+    // converted to degrees)
+    if(!kinematics.isInDegrees()) return nullptr;
```

Comment:
@aseth1 I changed the logic here; please verify this is what you intended. You originally had:

```cpp

    if(!kinematics.isInDegrees())
        if (pre40Model.getDocumentFileVersion() >= 30415) {
            throw Exception("updateKinematicsStorageForUpdatedModel has no updates "
                "to make because the model '" + pre40Model.getName() + "'is up-to-date.\n"
                "If input motion files were generated with this model version, "
                "nothing further must be done. Otherwise, provide the original model "
                "file used to generate the motion files and try again.");
        }
```

Reply:
:+1: yes returning makes most sense!